pkg/blueprint/firstboot_customizations.go

```diff
@@ -0,0 +1,173 @@
+type FirstbootScriptCustomization struct {
+	union json.RawMessage
+}
+
+// FirstbootCommonCustomization contains common fields for all firstboot customizations.
+type FirstbootCommonCustomization struct {
+	// Type of the firstboot customization. Supported values are:
+	// "custom", "satellite", and "aap".
+	Type string `json:"type,omitempty" toml:"type,omitempty"`
+
+	// Optional firstboot name. Must be unique within the blueprint and only
+	// alphanumeric characters with dashes and underscores are allowed.
+	Name string `json:"name,omitempty" toml:"name,omitempty"`
+
+	// Ignore errors when executing the firstboot script and continue with
+	// execution of the following firstboot scripts, if any. By default,
+	// firstboot scripts are executed in order and if one of them fails, the
+	// execution stops immediately.
+	IgnoreFailure bool `json:"ignore_failure,omitempty" toml:"ignore_failure,omitempty"`
+}
+
+// CustomFirstbootCustomization contains fields specific to custom firstboot
+// customizations.
+type CustomFirstbootCustomization struct {
+	FirstbootCommonCustomization
+
+	// Strings without shebang will be interpreted as shell scripts, otherwise
+	// the script will be executed using the shebang interpreter. Required if
+	// type is set to "custom".
+	Contents string `json:"contents" toml:"contents"`
+}
+
+// SatelliteFirstbootCustomization contains fields specific to satellite firstboot
+// customizations.
+type SatelliteFirstbootCustomization struct {
+	FirstbootCommonCustomization
+
+	// Optional CA certificate to enroll into the system before executing the
+	// firstboot script.
+	CACerts []string `json:"cacerts,omitempty" toml:"cacerts,omitempty"`
+
+	// Registration command as generated by the Satellite server. Required, if
+	// type is set to "satellite".
+	Command string `json:"command,omitempty" toml:"command,omitempty"`
+}
+
+// AAPFirstbootCustomization contains fields specific to AAP firstboot
+// customizations.
+type AAPFirstbootCustomization struct {
+	FirstbootCommonCustomization
+
+	// Optional CA certificate to enroll into the system before executing the
+	// firstboot script.
+	CACerts []string `json:"cacerts,omitempty" toml:"cacerts,omitempty"`
+
+	// Job template URL as generated by the AAP server. Required if type is set
+	// to "aap". Example URLs are
+	// https://aap.example.com/api/controller/v2/job_templates/9/callback/ or
+	// https://aap.example.com/api/v2/job_templates/9/callback/ depending on the
+	// AAP version.
+	JobTemplateURL string `json:"job_template_url,omitempty" toml:"job_template_url,omitempty"`
+
+	// The host config key. Required if type is set to "aap".
+	HostConfigKey string `json:"host_config_key,omitempty" toml:"host_config_key,omitempty"`
+}
+
+func (t FirstbootScriptCustomization) MarshalJSON() ([]byte, error) {
+	b, err := t.union.MarshalJSON()
+	return b, err
+}
+
+func (t *FirstbootScriptCustomization) UnmarshalJSON(b []byte) error {
+	err := t.union.UnmarshalJSON(b)
+	return err
+}
```

Comment:
So we kinda have the other TOML errors wrapped in case something goes wrong. I was just thinking if we should not try to wrap this one as well.
Just something like (very random string message), or maybe adding some more data about the byte array. :thinking: 
`	return fmt.Errorf("FirstbootScriptCustomization unmarshalling error: %w")`

Reply:
The marshaller error messages are usually pretty clear like "unexpected field named 'scriptx'" but sure, why not.